### PR TITLE
RCLC-Executor with new API and convenience functions

### PR DESCRIPTION
**Pull request by JanStaschulat**
This repository contains two packages: The rclc-package with an RCLC-Executor providing a new C-API based on the RCL-layer. It supports the standard rclcpp-Executor semantics and provides additional features for deterministic and real-time execution:
- sequential processing order
- triggered execution
- Logical Execution Time (LET) semantics for data synchronization.

Additionally, the rclc-package contains convenience functions to create RCL-objects to simplify setting up ROS2-applications in C.

Secondly, the package rclc_examples demonstrates how to use the RCLC-Executor and the convenience functions. 

For this Executor the [pull request 534](https://github.com/ros2/rcl/pull/534) on ros2/rcl was already created in Nov. 2019. This pull request resolves the comments of that PR and provides the new feature of triggered execution. After the [discussions](https://discourse.ros.org/t/ros2-real-time-working-group-online-meeting-5-nov-13-2019-7am-pdt-utc-7/11460) in the ROS2 Real-Time Working Group (tag: wg-real-time) it was decided to create a new pull request for ros2/rclc.

### CODE REVIEW

**Comment by JanStaschulat** [comment]
@wjwwood @dejanpan @fujitatomoya @jacobperron @Karsten1987 @ralph-lange 
I opened a new PR for the RCLC-Executor on ROS2/rclc. The source code integrates the reviewer comments of the previous [PR 534](https://github.com/ros2/rcl/pull/534). I made an effort, to motivate the features of the RCLC-Executor and provided several examples how to use it. 

Your feedback and comments are very much appreciated.

File: NOTICE [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Open Source Robotics Foundation, Inc.
+    William Woodall <william@osrfoundation.org>
+    Dirk Thomas <dthomas@osrfoundation.org>
+    Esteve Fernandez <esteve@osrfoundation.org>
+
+Robert Bosch GmbH
+    Jan Staschulat <jan.staschulat@de.bosch.com>
+    Ralph Lange <ralph.lange@de.bosch.com>
+
+Loïc Dauphin <astralien3000@yahoo.fr>
```
Comment (Karsten1987) [inline NOTICE]: missing newline

File: rclc/CMakeLists.txt [hunk 2]
```diff
@@ -1,56 +1,126 @@
...
+endif()
+
+# Default to C++14
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]: This package is a C-only package and doesn't require any C++ compiler, does it? If so, I don't think these parameters make much of an impact.

File: rclc/CMakeLists.txt [hunk 3]
```diff
@@ -1,56 +1,126 @@
...
+
+set(CMAKE_VERBOSE_MAKEFILE ON)
+
+#################################################
+# package dependencies
+#################################################
 find_package(ament_cmake_ros REQUIRED)
 find_package(rcl REQUIRED)
+find_package(rcutils REQUIRED)
+find_package(std_msgs REQUIRED)
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]: alphabetical sort

File: rclc/CMakeLists.txt [hunk 4]
```diff
@@ -1,56 +1,126 @@
...
-ament_export_include_directories(include)
-include_directories(include)
 
-if(NOT CMAKE_C_STANDARD)
-  set(CMAKE_C_STANDARD 11)
-endif()
+#################################################
+# include directories
+#################################################
+include_directories(include)
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]: which the below used `target_include_directories` I don't think this call is still needed.

File: rclc/CMakeLists.txt [hunk 5]
```diff
@@ -1,56 +1,126 @@
...
-ament_export_dependencies(
-  rcl
-  rmw
-  rosidl_generator_c
-)
+
+#################################################
+# install
+#################################################
+
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]:
```suggestion
```

File: rclc/CMakeLists.txt [hunk 6]
```diff
@@ -1,56 +1,126 @@
...
+install(TARGETS ${PROJECT_NAME} DESTINATION lib)
+
+# specific order: dependents before dependencies
 ament_export_include_directories(include)
 ament_export_libraries(${PROJECT_NAME})
 
+#################################################
+# gtest
+#################################################
+
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]:
```suggestion
```

File: rclc/CMakeLists.txt [hunk 7]
```diff
@@ -1,56 +1,126 @@
...
 if(BUILD_TESTING)
+  find_package(ament_cmake_gtest REQUIRED)
   find_package(ament_lint_auto REQUIRED)
+  set(ament_cmake_copyright_FOUND TRUE) # No copyright header check since link to NOTICE file is not recognized properly.
+  find_package(std_msgs REQUIRED)
+  find_package(geometry_msgs REQUIRED)
+  find_package(osrf_testing_tools_cpp REQUIRED)
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]: alpha sort

File: rclc/CMakeLists.txt [hunk 8]
```diff
@@ -1,56 +1,126 @@
...
+ament_export_dependencies(ament_cmake)
+ament_export_dependencies(rcl)
+ament_export_dependencies(rcutils)
+ament_export_dependencies(rosidl_generator_c)
 ament_package()
 
 install(
   DIRECTORY include/
   DESTINATION include
 )
```
Comment (Karsten1987) [inline rclc/CMakeLists.txt]: Isn't this redundant with line 69?

File: rclc/NOTICE [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+# If Individual's organization is copyright holder of her contributions add the
+# organization's name, optionally also the contributor's name:
+#
+#     Organization's name
+#         Individual's name <submission corporate email address>
+#
+# Please keep the list sorted.
+
+Robert Bosch GmbH
+    Jan Staschulat <jan.staschulat@de.bosch.com>
```
Comment (Karsten1987) [inline rclc/NOTICE]: missing newline

File: rclc/include/rclc/executor.h [hunk 10]
```diff
@@ -0,0 +1,387 @@
+// Copyright (c) 2019 - for information on the respective copyright owner
+// see the NOTICE file and/or the repository https://github.com/ros2/rclc.
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]: I think by putting these two lines just below the rest of the license, the linters should be okay with that.

File: rclc/include/rclc/executor.h [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+// Copyright 2014 Open Source Robotics Foundation, Inc.
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]: are you sure that this should be copyrighted by OSRF?

File: rclc/include/rclc/executor.h [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+  rclc_executor_semantics_t data_comm_semantics;
+} rclc_executor_t;
+
+/**
+ *  Return a rclc_executor_t struct with pointer members initialized to `NULL`
+ *  and member variables to 0.
+ */
+rclc_executor_t
+rclc_executor_get_zero_initialized_executor(void);
+/**
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]:
```suggestion

/**
```

File: rclc/include/rclc/executor.h [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param [inout] executor pointer to an initialized executor
+ * \param [in] valid semantics value as defined in enum type {@link rclc_executor_semantics_t}
+ * \return `RCL_RET_OK` if semantics was set successfully
+ * \return `RCL_RET_INVALID_ARGUMENT` if \p executor is a null pointer
+ */
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]:
```suggestion
```

File: rclc/include/rclc/executor.h [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+ *
+ * \param [inout] executor pointer to initialized executor
+ * \return `RCL_RET_OK` if spin operation was successful
+ * \return `RCL_RET_INVALID_ARGUMENT` if executor is a null pointer
+ * \return `RCL_RET_ERROR` if any other error occured
+ */
+rcl_ret_t
+rclc_executor_spin(rclc_executor_t * executor);
+
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]:
```suggestion
```

File: rclc/include/rclc/executor.h [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+ * \return `RCL_RET_OK` if spin operation was successful
+ * \return `RCL_RET_INVALID_ARGUMENT` if executor is a null pointer
+ * \return `RCL_RET_ERROR` if any other error occured
+ */
+rcl_ret_t
+rclc_executor_spin_period(
+  rclc_executor_t * executor,
+  const uint64_t period);
+
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]:
```suggestion
```

File: rclc/include/rclc/executor.h [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+ rclc_executor_spin_period_ implements one iteration and the function
+ rclc_executor_spin_period implements the endless while-loop. The unit test covers only
+ rclc_executor_spin_period_.
+*/
+rcl_ret_t
+rclc_executor_spin_one_period(
+  rclc_executor_t * executor,
+  const uint64_t period);
+
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]:
```suggestion
```

File: rclc/include/rclc/executor.h [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+/*
+ The reason for splitting this function up, is to be able to write a unit test.
+ The spin_period is an endless loop, therefore it is not possible to stop after x iterations. The function
+ rclc_executor_spin_period_ implements one iteration and the function
+ rclc_executor_spin_period implements the endless while-loop. The unit test covers only
+ rclc_executor_spin_period_.
+*/
```
Comment (Karsten1987) [inline rclc/include/rclc/executor.h]: I am not sure if the style (different from the style above) will be picked up from doxygen like this.

Comment (Karsten1987) [reply rclc/include/rclc/executor.h]: same below

File: rclc/include/rclc/executor_handle.h [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+  /// Total number of timers
+  size_t number_of_timers;
+  /// Total number of clients
+  size_t number_of_clients;
+  /// Total number of services
+  size_t number_of_services;
+  /// Total number of events
+  size_t number_of_events;
+} rclc_executor_handle_counters_t;
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor_handle.h]:
```suggestion
```

File: rclc/include/rclc/executor_handle.h [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] handle preallocated rclc_executor_handle_t
+ * \return `RCL_RET_OK` if the handle was printed successfully
+ * \return `RCL_RET_INVALID_ARGUMENT` if \p h is a null pointer
+ */
+rcl_ret_t
+rclc_executor_handle_print(rclc_executor_handle_t * handle);
+
```
Comment (Karsten1987) [inline rclc/include/rclc/executor_handle.h]:
```suggestion
```

File: rclc/include/rclc/init.h [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+ * \return `RCL_RET_INVALID_ARGUMENT` if any null pointer as argument
+ * \return `RCL_RET_ERROR` in case of failure
+ */
+rcl_ret_t
+rclc_support_init(
+  rclc_support_t * support,
+  int argc,
+  char const * const * argv,
+  rcl_allocator_t * allocator);
+
```
Comment (Karsten1987) [inline rclc/include/rclc/init.h]:
```suggestion
```

File: rclc/include/rclc/node.h [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+#ifndef RCLC__NODE_H_
+#define RCLC__NODE_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+#include <rcl/node.h>
+#include <rclc/types.h>
+#include <rclc/init.h>
```
Comment (Karsten1987) [inline rclc/include/rclc/node.h]:
```suggestion

#include <rclc/init.h>
```

File: rclc/include/rclc/rclc.h [hunk 22]
```diff
@@ -23,77 +25,12 @@ extern "C"
...
-
-/// Causes the executor to process callbacks generated by ROS Nodes.
-rclc_ret_t
-rclc_executor_spin();
+#include "rclc/types.h"
+#include "rclc/init.h"
+#include "rclc/node.h"
+#include "rclc/publisher.h"
+#include "rclc/subscription.h"
+#include "rclc/timer.h"
```
Comment (Karsten1987) [inline rclc/include/rclc/rclc.h]: alpha sort

**Comment by JanStaschulat** [comment]
@Karsten1987 Regarding linters:
All linters pass without errors: 

I run with alias in .bashrc:
alias colcon_test="colcon test --base-path=src --event-handlers console_direct+ --pytest-args \" --capture=no\""

Output of colcon_test --packages-select rclc
...
100% tests passed, 0 tests failed out of 6

Label Time Summary:
cppcheck      =   0.45 sec*proc (1 test)
cpplint       =   0.94 sec*proc (1 test)
gtest         =  10.93 sec*proc (1 test)
lint_cmake    =   0.26 sec*proc (1 test)
linter        =   4.45 sec*proc (5 tests)
uncrustify    =   0.37 sec*proc (1 test)
xmllint       =   2.44 sec*proc (1 test)

Total Test time (real) =  15.39 sec
Finished <<< rclc [15.5s]

File: rclc/CMakeLists.txt (hunk 2 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: The unit tests are written in C++.

File: rclc/include/rclc/executor.h (hunk 10 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: Hmm, I don't get any errors from the linters. Which linters creates an error message here?

File: rclc/include/rclc/executor.h (hunk 11 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: Yes. OSRF started the development of rclc.

File: NOTICE (hunk 1 above)
Comment (JanStaschulat) [reply NOTICE]: Fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 3 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: Fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 4 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: Fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 5 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: Fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor_handle.h (hunk 18 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor_handle.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor_handle.h (hunk 19 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor_handle.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/init.h (hunk 20 above)
Comment (JanStaschulat) [reply rclc/include/rclc/init.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/node.h (hunk 21 above)
Comment (JanStaschulat) [reply rclc/include/rclc/node.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/rclc.h (hunk 22 above)
Comment (JanStaschulat) [reply rclc/include/rclc/rclc.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 6 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 7 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/CMakeLists.txt (hunk 8 above)
Comment (JanStaschulat) [reply rclc/CMakeLists.txt]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/NOTICE (hunk 9 above)
Comment (JanStaschulat) [reply rclc/NOTICE]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 12 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 13 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 14 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 15 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 16 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

File: rclc/include/rclc/executor.h (hunk 17 above)
Comment (JanStaschulat) [reply rclc/include/rclc/executor.h]: fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b

**Comment by JanStaschulat** [comment]
@wjwwood @dejanpan @fujitatomoya @jacobperron @Karsten1987

**Comment by carlossvg** [comment]
@JanStaschulat Just some minor comments from my side:

1. I would use a more descriptive name than RCLCPP_EXECUTOR. I understand that currently there is not a better choice to describe its semantics. 
2. It would be interesting to add an example specifically showing  LET semantics benefits over rclcpp in terms of data determinism. For example showing some output differences when setting  first "LET" semantics and then "RCLCPP_EXECUTOR" semantics. 
3. Also, could your provide some insights about how LET semantics would work in a multi-threaded context?
4. Some typo fixes in the README: https://github.com/micro-ROS/rclc/pull/17

**Comment by JanStaschulat** [comment]
@carlossvg Thanks for your remarks.
1. Yes, it was also difficult for me to come up with a good name. What do you think about the following reasoning: Lets look at the characteristics of the different semantics: 
    - LET: reading all available data from DDS at the beginning of spin_some(), processing all callbacks, publishing all data at the end (of the period) at once.
   - rclc Executor: checking for new data at DDS queue, fetch data of topic just before the corresponding callback is executed, data is published within the callback code.
So I would propose to IMMEDIATE as a more descriptive name for the semantics of the rclc executor, as 1) data is fetched immediately before the corresponding callback is called and 2) data is immediately published, when publish() is called. (Synonym could also be INSTANTLY). 
@wjwwood  What do you think? Do you prefer any other desciptive name for the current rclc semantics?

2. There are two test cases: semantics_LET and semantics_RCLCPP showing the different semantics. The setup is: 
  // configuration
  // - one publisher
  //    - publishes integer topic X
  // - subscriber A
  //    - subscribes to X
  //    - publishes also on topic X
  // - subscriber B
  //     - subscribes to X with DDS quality of service option: last-is-best
  //     - evaluates data of topic X and copies it to global variable result_X
  // - evaluation in test case, what the value of result_X is.
  //
  // test setup
  //  - publisher: publishes X=1
  //  - sleep(100ms)
  //  - spin_some()
  //  - subscriber A : publishes X=2
  //  - sleep(100ms)
  //  - subscriber B :
  // expected test result for semantics RCLCPP:
  //   => subscriber B receives X=2 (because it takes most recent data)
  // expected test result for semantics LET:
  //   => subscriber B receives X=1 (because data is taken from DDS at the start of spin_some())

Do you think to have this as an explicit example in the rclc_examples directory?

3. The current implementation supports only the single threaded use-case. In a next step, we want to support the multi-threaded use-case as well. That means to extend the API of the RCLC Executor with a rclc_publish() function and some additional messages for synchronization of processes

3.1 rclc_publish() function:

To be used in the callback like this:
callback(){
    // some code
    rclc_publish(&sub, &msg); 
}

This function will just add the topic and the message into a queue, which is managed by the RCLC Executor. At initialization, the user must specify the length of this queue and the period, when all the collected messages shall be published. 

rclc_executor {
 unsigned int max_pub_size;
 void * pub_queue;
}
 rclc_executor_configure_LET_publisher(max_size, period_ms);
 rclc_executor__LET_publish(); 
 rclc_executor_publish(rcl_subscription_t *; void *);


Please refer to Figure 3 in [rclc documentation](https://github.com/micro-ROS/rclc/tree/master/rclc) 
The assumption is that each process will run in a different thread and has a fixed period. That is, every process is periodically executed, in embedded context that is typically 1ms, 5ms, 10ms, 20ms, 100ms time slices. 

By the way, this time-driven paradigm is also applicable for robotics applications, as the sensor (e.g. laser) publishes data periodically and drives with this the entire (data-driven) pipeline.

- At initialization of the LET-Executor, the publish-queue is configured with the maximum number of elements and the period, how often to publish. 
- Within a callback the rclc_publish() function will put the data into a pub_queue. 
- Every 'period' all data in the pub_queue will be published by rclc_executor_LET_publish()

This completes the semantics for LET
- reading data at the beginning of the period (already implemented)
- publishing data at the end of the period

3.2 Synchronization between threads (processes)
- There needs to be a synchronization mechanism between the different threads to make sure that publishing is atomic.  e.g. that publishing the data from the  1ms process occurs completely before any other process reads this data. 
- One idea is to use a dedicated "pub-ready" message which then is used as trigger condition for the other processes. 
- this needs to be worked out in more detail

4. merged pull request.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-02-20/12901/1

**Comment by JanStaschulat** [comment]
@carlossvg @wjwwood @dejanpan @Karsten1987 What is the status of your discussion for merging this pull request?

**Comment by wjwwood** [comment]
I'm happy to get out of you guys way and let you merge things on this repository without my review. I simply don't have the bandwidth to provide reviews for another client library on top of rclcpp and rclpy, sorry. :(

If you guys need permissions or someone else on @ros2/team wants to help with reviews let me know.

**Comment by Karsten1987** [comment]
I think it would make sense to transfer the ownership of this repo to the realtime working group similar to how rosbag2 and friends are managed and reviewed by the tooling working group.
We can create a github team which has the appropriate rights to review and merge PRs on this repo.

Thoughts?

**Comment by ralph-lange** [comment]
We are happy to take over the ownership, possibly together with a partner not directly involved the micro-ROS development to have a second (unbiased) view on the API.
@wjwwood, @carlossvg Is there anybody from the Real-Time WG to volunteer? Otherwise, I would start with a PR to transfer the ownership to Bosch and eProsima for the time being, to be able to review our PRs mutually.

**Comment by carlossvg** [comment]
@ralph-lange It makes sense to me that Bosch and eProsima take over the ownership as they are the main entities involved. Anyway, we can comment this in the next RTWG meeting in case there is someone else interested.

**Comment by wjwwood** [comment]
We can leave the repository in this organization. I can just setup a team for you guys so you can administrate it yourself and add whoever you want to it. I can also help you if you'd like to move it to another org, but I think leaving it here is best for visibility and discovery personally.

Just let me know who and I'll set it up!

**Comment by ralph-lange** [comment]
If no one else from the Real-time WG is eager to take ownership of rclc in today's meeting, then please make @JanStaschulat and @julibert owners of it.

**Comment by wjwwood** [comment]
@ralph-lange I've added @JanStaschulat and @julibert to the @ros2/rclc team. I can add others too, if needed. If you all would let me know when you've accepted the invitations I can update your permissions afterwards.

**Comment by JanStaschulat** [comment]
@wjwwood Accepted the invitation.

**Comment by ralph-lange** [comment]
@wjwwood   Jan and Julian both accepted the invitation, you may make them owners now.

**Comment by wjwwood** [comment]
You both now have the ⚡️! Thanks for helping with this package.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-03-18/13313/1

**Pull request merged**